### PR TITLE
Fix race condition in web_server scheduler on ESP32

### PR DESCRIPTION
**Pull request by tomaszduda23**
# What does this implement/fix?
Race condition in scheduler

Web component schedule things from async_tcp thread which leads to race condition.
It puts things to queue and schedule them from main thread.
## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>
https://github.com/esphome/issues/issues/2916
https://github.com/esphome/issues/issues/2632
https://github.com/esphome/esphome/pull/3943
https://github.com/esphome/esphome/pull/3917
**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [x] ESP32-S2

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/web_server/web_server.cpp [hunk 1]
```diff
@@ -120,7 +127,33 @@ void WebServer::setup() {
...
+#ifdef USE_ESP32
+#if 0
+  if (xSemaphoreTake(this->to_schedule_lock_, 0L)) {
+    for (auto &it : this->to_schedule_) {
+      this->defer(std::move(it));
+    }
+    this->to_schedule_.clear();
+    xSemaphoreGive(this->to_schedule_lock_);
+  }
+#else
```
Comment (jesserockz) [inline esphome/components/web_server/web_server.cpp]: No point keeping this if it will never execute

**Comment by ssieb** [comment]
Is this an issue specific to ESP32 and not applicable to ESP8266?

**Comment by tomaszduda23** [comment]
> Is this an issue specific to ESP32 and not applicable to ESP8266?

ESP8266 uses different library ottowinter/ESPAsyncTCP-esphome. I took quick look and cannot see any task there. @OttoWinter are you able to confirm that? I have no ESP8266 to test.

**Comment by tomaszduda23** [comment]
https://github.com/esphome/issues/issues/3780

**Review by jesserockz — APPROVED** [review]

**Pull request merged**